include/revng/EarlyFunctionAnalysis/FunctionMetadataCache.h

```diff
@@ -48,45 +48,86 @@ extractFunctionMetadata(const llvm::BasicBlock *BB) {
 
 } // namespace detail
 
-class FunctionMetadataCache {
-private:
-  std::map<const llvm::Value *, efa::FunctionMetadata> FunctionCache;
+template<typename T>
+concept FunctionMetadataCacheTraits = requires {
```

Comment:
Add documentation about why we needs this stuff

Reply:
Do you mean why the *concept* is needed? The answer is that it isn't really. It just gives errors earlier if you pass a traits class that doesn't provide the necessary interface. I only added it because @pfez asked for it.

Reply:
Nono, I mean what's the use case for having a trait. Why not just a regular class? The answer is: we need to implement it for both LLVM IR and MLIR.

Reply:
I added a comment explaining it.